file.c

```diff
@@ -276,12 +348,103 @@ int kibosh_release(const char *path UNUSED, struct fuse_file_info *info)
                  struct fuse_file_info *info)
 {
-    int ret;
+    int ret, fault;
     uint32_t uid UNUSED;
     struct kibosh_file *file = (struct kibosh_file*)(uintptr_t)info->fh;
+    struct kibosh_fs *fs = fuse_get_context()->private_data;
+
+    if (file->type == KIBOSH_FILE_TYPE_NORMAL) {
+        fault = kibosh_fs_check_write_fault(fs, file->path);
+
+        // inject write_corrupt fault
+        if (fault >= CORRUPT_ZERO) {
+            double fraction = 0.0;
+            char *file_type = NULL;
+            struct kibosh_fault_base **iter;
+            struct kibosh_fault_write_corrupt *corrupt_fault;
+            for (iter = fs->faults->list; *iter; iter++) {
+                if (strcmp((*iter)->type, KIBOSH_FAULT_TYPE_WRITE_CORRUPT) == 0) {
+                    corrupt_fault = (struct kibosh_fault_write_corrupt*)(*iter);
+                    fraction = corrupt_fault->fraction;
+                    file_type = corrupt_fault->file_type;
+                }
+            }
+
+            int pos = (int) ((1.0 - fraction) * size);
+            int buf_size = (int) size;
+
+            switch(fault) {
+                case CORRUPT_RAND:
+                    for (int i=0; i < buf_size; ++i) {
+                        if (RAND_FRAC <= fraction) {
+                            memset(buf+i, (int) round(RAND_FRAC * 255.0), 1);
+                        }
+                    }
+                    break;
+
+                case CORRUPT_ZERO:
+                    for (int i=0; i < buf_size; ++i) {
+                        if (RAND_FRAC <= fraction) {
+                            memset(buf+i, 0, 1);
+                        }
+                    }
+                    break;
+
+                case CORRUPT_RAND_SEQ:
+                    for (; pos < buf_size; ++pos) {
+                        memset(buf+pos, (int) round(RAND_FRAC * 255.0), 1);
+                    }
+                    break;
+
+                case CORRUPT_ZERO_SEQ:
+                    memset(buf+pos, 0, buf_size - pos);
+                    break;
+
+                case CORRUPT_DROP:
+                    size = pos;
+                    break;
+
+                // CORRUPT_ZERO
+                default:
+                    memset(buf+pos, 0, buf_size - pos);
+            }
+            ret = pwrite(file->fd, buf, size, offset);
+            uid = fuse_get_context()->uid;
+            DEBUG("kibosh_write(file->path=%s, size=%zd, offset=%" PRId64", uid=%"PRId32") "
+                                      "= write_corrupt{mode=%d, fraction=%g, file_type=%s, ret=%d}\n",
+                                      file->path, size, (int64_t)offset, uid, fault, fraction, file_type, ret);
+            char buf_write[ret + 1];
+            for (int i=0; i<ret; ++i) {
+                if (buf[i] == 0) {
+                    buf_write[i] = '?';
+                } else {
+                    buf_write[i] = buf[i];
+                }
+            }
+            buf_write[ret] = '\0';
```

Comment:
We are using `buf_write` only when we want to debug. Is it possible to check the DEBUG macro is enabled before we do this copy operation. Also, this debug message makes more sense when we are writing zeroes as part of corruption. When we write a random integer, it is not useful.
User may supply a flag when they need to see the buffer data. Most of the times, it should be ok to tell that we corrupted the buffer.

Similar comment for the `buf_read` buffer at read function

Reply:
A flag was added to corruption faults to let user decide whether to keep the corrupted data for reference.